Delta Migration: Handle delta lake's automatic log clean and VACUUM

JonasJ-ap opened the pull request:
## Issue Fixed
This PR fixes #6781 .

The automatic log clean whenever a new checkpoint is created and the manual data file clean by `VACUUM` sometimes lead to inconsistency between the first constructable delta lake version and the earliest delta lake version in the table. The inconsistency will cause file not found errors during the migration.

This PR fix the issue by forcing the snapshot delta lake action to start from the first constructable delta lake version instead of the earliest version.

## Test
Added Integration tests to mock the file clean scenario.

jackye1995 commented on delta-lake/src/integration/java/org/apache/iceberg/delta/TestSnapshotDeltaLakeTable.java:
```diff
@@ -113,6 +124,8 @@ public static Object[][] parameters() {
   @Rule public TemporaryFolder temp4 = new TemporaryFolder();
   @Rule public TemporaryFolder temp5 = new TemporaryFolder();
+  @Rule public TemporaryFolder temp6 = new TemporaryFolder();
+  @Rule public TemporaryFolder temp7 = new TemporaryFolder();
```
can we try to avoid creating more temp folders? We should be able to reuse the same one for each test

JonasJ-ap replied:
Thank you for your suggestion. I misunderstood the logic of this temp folder. I revised the code to reuse the same one for each test

jackye1995 reviewed (commented):
thanks for the fix, overall looks good to me, just nit for test

jackye1995 commented:
This seems needed as a fix for Delta conversion to work properly, because most users run VACUUM against the table. Mark it as required for 1.2 release

amogh-jahagirdar commented on delta-lake/src/main/java/org/apache/iceberg/delta/BaseSnapshotDeltaLakeTableAction.java:
```diff
@@ -213,6 +218,52 @@ private PartitionSpec getPartitionSpecFromDeltaSnapshot(Schema schema) {
+        appendFiles.commit();
+
+        return constructableStartVersion;
+      } catch (NotFoundException | IllegalArgumentException | DeltaStandaloneException e) {
```
Just want to confirm we don't need to explicitly handle `IOException` here for when the file isn't found, we'll get a `NotFoundException`?

JonasJ-ap replied:
Thank you for pointing this out. The `HadoopFileIO` re-throw `FileNotFoundExceptiion` as `NotFoundException` and general `IOException` as `RuntimeIOException`.

https://github.com/apache/iceberg/blob/b5a31a14d56c1ee24bad87e1ac7f119d638ee320/core/src/main/java/org/apache/iceberg/hadoop/HadoopInputFile.java#L159-L170

In this case, I think we only need to handle the `NotFoundException` since `VACUUM` may delete the data file. I've added a code block in `buildDataFileFromAction` to explicitly check file existence and throw `NotFoundException` if necessary.

Please let me know if you have any other concern here.

amogh-jahagirdar commented on delta-lake/src/main/java/org/apache/iceberg/delta/BaseSnapshotDeltaLakeTableAction.java:
```diff
@@ -213,6 +218,52 @@ private PartitionSpec getPartitionSpecFromDeltaSnapshot(Schema schema) {
+      try {
+        List<AddFile> initDataFiles =
+            deltaLog.getSnapshotForVersionAsOf(constructableStartVersion).getAllFiles();
+        List<DataFile> filesToAdd = Lists.newArrayList();
```
Actually nvm, this is not possible, this will end up being the files for a single initial snapshot.

JonasJ-ap commented on delta-lake/src/main/java/org/apache/iceberg/delta/BaseSnapshotDeltaLakeTableAction.java:
```diff
@@ -310,6 +351,11 @@ private DataFile buildDataFileFromAction(Action action, Table table) {
+    if (!file.exists()) {
+      throw new NotFoundException(
+          "The file %s does not exist in the Delta Lake table at %s",
+          fullFilePath, deltaTableLocation);
```
I think checking file existence explicitly here can make the code logic more clear and have more reasonable error message

jackye1995 replied:
nit: we can be more precise about the message, something like "File %s is referenced in Delta lake metadata but cannot be found in storage"

jackye1995 commented on delta-lake/src/integration/java/org/apache/iceberg/delta/TestSnapshotDeltaLakeTable.java:
```diff
@@ -122,28 +129,36 @@ public TestSnapshotDeltaLakeTable(
     spark.sql(String.format("DROP TABLE IF EXISTS %s", unpartitionedIdentifier));
     spark.sql(String.format("DROP TABLE IF EXISTS %s", externalDataFilesIdentifier));
     spark.sql(String.format("DROP TABLE IF EXISTS %s", typeTestIdentifier));
+    spark.sql(String.format("DROP TABLE IF EXISTS %s", vacuumTestIdentifier));
```
Instead of creating one subfolder for each table, I think we can just create tabes within the test, and do a `DROP DATABASE CASCADE` call in the end to remove all the tables after all tests

JonasJ-ap replied:
Thank you for your suggestions. I refactored integration tests to create tables within the test and remove redundant sql commands.

jackye1995 commented on delta-lake/src/integration/java/org/apache/iceberg/delta/TestSnapshotDeltaLakeTable.java:
```diff
@@ -65,26 +68,23 @@ public class TestSnapshotDeltaLakeTable extends SparkDeltaLakeSnapshotTestBase {
-  private final String snapshotExternalDataFilesTableName = "iceberg_external_data_files_table";
-  private final String snapshotNewTableLocationTableName = "iceberg_new_table_location_table";
-  private final String snapshotAdditionalPropertiesTableName =
+  private static final String partitionedTableName = "partitioned_table";
```
as best practice for testing, we can just hard code these names to individual tests, no need to make them static variables

jackye1995 commented on delta-lake/src/integration/java/org/apache/iceberg/delta/TestSnapshotDeltaLakeTable.java:
```diff
@@ -121,91 +117,28 @@ public TestSnapshotDeltaLakeTable(
-            "DROP TABLE IF EXISTS %s", destName(icebergCatalogName, snapshotTypeTestTableName)));
-
-    spark.sql(String.format("DROP DATABASE IF EXISTS %s", NAMESPACE));
+    spark.sql(String.format("DROP DATABASE IF EXISTS %s CASCADE", NAMESPACE));
```
this should be in @AfterClass hook instead of @After

JonasJ-ap commented:
Thanks @jackye1995, @amogh-jahagirdar for reviewing this. I added the file existence check and refactored the integration tests based on your suggestions. @amogh-jahagirdar May I ask if you could review this PR again? Thank you in advance for your help!

jackye1995 commented on delta-lake/src/integration/java/org/apache/iceberg/delta/TestSnapshotDeltaLakeTable.java:
```diff
@@ -454,7 +472,7 @@ private Dataset<Row> typeTestDataFrame() {
   }
 
-  private Dataset<Row> nestedDataFrame() {
+  private static Dataset<Row> nestedDataFrame() {
```
this can be a static variable instead of a static method

JonasJ-ap replied:
Thank you for your suggestion. I moved the dataframe initialization code blocks to `BeforeClass`.

jackye1995 reviewed (approved):
looks good to me!

jackye1995 commented:
Merging, thanks for the fix @JonasJ-ap , and thanks for the review @yyanyy and @amogh-jahagirdar !
